### DISCUSSION TITLE
curl interactions

### DISCUSSION THREAD

**Discussion by ChristiaanEmpowerBI** [MicroStrategy/rest-api-docs#36] [Q&A] [answered]
There is something that I do not seem to understand. 

I have a MicroStrategy Server running on my local machine, running Windows 10, and MicroStrategy 2021 Update 11. 

I connect to the REST API using a curl command and sending on json data. 

curl -i -X POST http://my_pc:8080/MicroStrategyLibrary/api/auth/login -H "accept: application/json" -H "Content-Type: application/json" -d "{\"username\":\"REST_API_User\",\"password\":\"rest_password\",\"loginMode\":1}" 

Then I get a response:
HTTP/1.1 204
Set-Cookie: JSESSIONID=85F6E15A8B2A881D51BC2AD733B540D4; Path=/MicroStrategyLibrary; HttpOnly
Set-Cookie: iSession=874jjcr94644beqhurppqdu7ng; Path=/MicroStrategyLibrary; HttpOnly
X-MSTR-AuthToken: my_token_supplied_by_the_server
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
Date: Mon, 02 Oct 2023 14:22:27 GMT

So I am under the impression that I am successfully authenticated, and have been supplied a Authentication Token, and can now interact with the REST API using this token. 

So I send my second curl command:
curl -i -X GET http://my_pc:8080/MicroStrategyLibrary/api/projects -H "accept: application/json" -H "X-MSTR-AuthToken: my_token_supplied_by_the_server"

And I get the response:
HTTP/1.1 401
Set-Cookie: JSESSIONID=2D4A23C98935C5A46AFC8955CC238006; Path=/MicroStrategyLibrary; HttpOnly
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Date: Mon, 02 Oct 2023 14:25:49 GMT

{"code":"ERR009","message":"The user's session has expired, please reauthenticate","ticketId":"5ee91f76906a48cb849fc0f1622f6fcc"}

looking at the documentation
https://demo.microstrategy.com/MicroStrategyLibrary/api-docs/index.html#/Authentication/sessionSessionIdPut
this message seems to be a standard error message so I do not think this error message is helpful nor true, but do not know how to troubleshoot it more. 

![image](https://github.com/MicroStrategy/rest-api-docs/assets/139224196/d72fe708-ec2e-475d-ad5d-ca08b55b6973)

The MicroStrategy error logs in 
D:\Program Files (x86)\Common Files\MicroStrategy\Log
are also of no help. 

Anyone have any ideas for me, please?

**Comment by ChristiaanEmpowerBI** [comment]
I figured it out watching a video of Daniel Zumbaugh, you need to add the JSESSIONID info too when using curl.

**Reply by Dholmblad** [reply, marked as answer]
@ChristiaanEmpowerBI Be aware other cookies may be required depending on infrastructure configuration, such as load balancer. This is due to sticky sessions being required configuration for horizontally scaled MicroStrategyLibrary applications. 

See documentation for reference:https://www2.microstrategy.com/producthelp/2020/RESTSDK/Content/topics/REST_API/REST_API_Workflow_Authentication.htm
